src/net/neighbors/walk.rs

```diff
@@ -0,0 +1,2003 @@
+                       local_peer_str,
+                       &neighbor_from_handshake.addr,
+                       &self.cur_neighbor.addr);
+
+            return Err(net_error::PeerNotConnected);
+        };
+
+        debug!(
+            "{}: Connected with {:?}",
+            local_peer_str, &self.cur_neighbor.addr
+        );
+
+        let cur_neighbor = self.cur_neighbor.clone();
+        let new_cur_neighbor =
+            self.neighbor_db
+                .update_neighbor(network, cur_neighbor, Some(data), db_data)?;
+        self.cur_neighbor = new_cur_neighbor;
+
+        self.new_frontier
+            .insert(self.cur_neighbor.addr.clone(), self.cur_neighbor.clone());
+        self.neighbor_from_handshake = neighbor_from_handshake.addr;
+
+        Ok(self.cur_neighbor.clone())
+    }
+
+    /// Finish handshaking with our current neighbor, thereby ensuring that it is connected
+    /// Returns true if we finished talking to the neighbor
+    /// Returns false if not
+    pub fn handshake_try_finish(&mut self, network: &mut PeerNetwork) -> Result<bool, net_error> {
+        assert!(self.state == NeighborWalkState::HandshakeFinish);
+        if self.comms.count_inflight() == 0 {
+            // can't proceed
+            debug!("{:?}: No messages inflight", network.get_local_peer());
+            return Err(net_error::PeerNotConnected);
+        }
+
+        let message = if let Some((_, message)) = self.comms.collect_replies(network).pop() {
+            message
+        } else {
+            // try again later
+            return Ok(false);
+        };
+
+        match message.payload {
+            StacksMessageType::HandshakeAccept(ref data) => {
+                debug!(
+                    "{:?}: received HandshakeAccept from {} {:?}: {:?}",
+                    network.get_local_peer(),
+                    if self.walk_outbound {
+                        "outbound"
+                    } else {
+                        "inbound"
+                    },
+                    &message.to_neighbor_key(&data.handshake.addrbytes, data.handshake.port),
+                    &data.handshake
+                );
+
+                self.handle_handshake_accept(network, &message.preamble, data, None)?;
+            }
+            StacksMessageType::StackerDBHandshakeAccept(ref data, ref db_data) => {
+                debug!(
+                    "{:?}: received StackerDBHandshakeAccept from {} {:?}: {:?}, {:?}",
+                    network.get_local_peer(),
+                    if self.walk_outbound {
+                        "outbound"
+                    } else {
+                        "inbound"
+                    },
+                    &message.to_neighbor_key(&data.handshake.addrbytes, data.handshake.port),
+                    &data.handshake,
+                    db_data
+                );
+
+                self.handle_handshake_accept(network, &message.preamble, data, Some(db_data))?;
+            }
```

Comment:
Thes `handshake_accepts` seem like they could be deduped by doing something like:

```
let (handshake_data, stacker_db_data) = match message.payload {
    HandshakeAccept(ref data) => (data, None),
    StackerDBHandshakeAccept(ref data, ref db_data) => (data, Some(db_data)),
     ...
};
```